design-proposals/klip-7-connect-integration.md

```diff
@@ -0,0 +1,232 @@
+           
+## Motivation and background
+
+Databases provide three functions:
+- Data ingress (i.e. inserting data into the database)
+- Data transform (i.e. deriving new data from old)
+- Data egress (i.e. querying and/or ETL)
+
+In order to develop KSQL as the streaming database, we must provide a complete story around each of
+these functionality. Today, KSQL is a compelling tool to accomplish data transformations but relies
+on external mechanisms to import and export data. This KLIP lays out foundations for an integration 
+with Kafka Connect, a system that provides powerful primitives for data ingress and egress for 
+Apache Kafka.
+
+### Motivating Example
+
+This document will extend the canonical KSQL example of joining pageviews with a user table, but in
+this document we will assume that the user table is stored in a JDBC compatible database, as opposed
+to a Kafka changelog topic. After implementing this KLIP, it should be possible for a KSQL user to
+perform a stream-table join on these two without leaving KSQL's console.
+
+## Scope
+
+This KLIP will:
+- define a SQL-like language for basic connector management
+- provide a guiding principles and high level architecture for a connect-ksql integration
+
+This KLIP will _not_:
+- propose improvements to Kafka Connect such as easing the configuration burden, but will discuss
+  how to leverage future Connect improvements
+
+## Value/Return
+
+This proposal targets _usability_ improvements for KSQL and simplifies bootstrapping KSQL for users
+that may already have some, but not all, of their data in Kafka. As with the motivating example,
+KSQL users will be able to easily model external systems within KSQL. The target user has already
+played around with KSQL and is working on prototyping their first real application with existing
+data.
+
+## Design
+
+### Architecture
+
+There are three ways to integrate KSQL with Connect:
+1. Run them separately and supply syntax to communicate between them
+1. Embed Connect into KSQL, managing individual connect workers as part of the KSQL ecosystem
+1. Package Connect and KSQL into a single JVM, but just spin them up at the top level
+
+There is some consensus that at scale KSQL and Connect should run as separate components. This
+means that the API that we design will not be able to leverage any "embedded-ness" of Connect,
+nullifying most of the benefits of the second approach. 
+
+There are still, however, benefits to packaging connect and KSQL in the same JVM for a seamless
+experience for new users. Namely, new users will only need to download a single package, spin up
+a single process and maintain only a single application. There are some (solvable) concerns with
+this approach, but for the reason outlined above, the syntax will be developed with the assumption
+that Connect is running as a separate component.
+
+**The recommendation is to support 1 & 3, tabling 2 unless a need for it comes in the future**
```

Comment:
Is the proposal for the first version just (1) or support both (1) and (3)?

Reply:
both 1&3